### PR TITLE
[REEF-1080]  Make local temp file configurable in FileSystemInputPart…

### PR DESCRIPTION
**Pull request by jwang98052**
…ition

*Add ITempFileCreator interface and WorkingDirectoryTempFileCreator similar as Java did without creating File object as it doesn't make sense in C#
*Add WorkingDirectoryTempFileConfigModule and WorkingDirectoryTempFileConfigProvider
*Add ITempFileCreator as a parameter of FileSystemInputPartition injectable constructor
*add and modifies test cases
JIRA: [REEF-1080](https://issues.apache.org/jira/browse/REEF-1080)

This closes #

### CODE REVIEW

File: lang/cs/Org.Apache.REEF.Common/Io/TempFileFolder.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+using Org.Apache.REEF.Tang.Annotations;
+
+namespace Org.Apache.REEF.Common.Io
+{
+    /// <summary>
+    /// Folder for local temp file
+    /// </summary>
+    [NamedParameter("TempFileFolder", "TempFileFolder", @".\reef\tmp\")]
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Io/TempFileFolder.cs]: Please use forward slashes. Eventually, we'll want to run REEF.NET on linux :-)

File: lang/cs/Org.Apache.REEF.Common/Io/WorkingDirectoryTempFileCreator.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        [Inject]
+        private WorkingDirectoryTempFileCreator(
+            [Parameter(typeof(TempFileFolder))] string tempFileFolder)
+        {
+            _tempFileFolder = tempFileFolder;
+            if (!Directory.Exists(tempFileFolder))
+            {
+                Directory.CreateDirectory(tempFileFolder);
+            }
+            Logger.Log(Level.Info, string.Format(CultureInfo.CurrentCulture, "Temp directory: {0} is craeted", Path.GetFullPath(tempFileFolder)));
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Io/WorkingDirectoryTempFileCreator.cs]: Typo

File: lang/cs/Org.Apache.REEF.Common/Io/WorkingDirectoryTempFileCreator.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        }
+
+        /// <summary>
+        /// returns teh temp file folder
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Io/WorkingDirectoryTempFileCreator.cs]: Typo

File: lang/cs/Org.Apache.REEF.Common/Io/WorkingDirectoryTempFileCreator.cs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+using System.Globalization;
+using System.IO;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Utilities.Logging;
+
+namespace Org.Apache.REEF.Common.Io
+{
+    /// <summary>
+    /// This class wraps temp file folder and temp file folder creation
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Io/WorkingDirectoryTempFileCreator.cs]: Too unspecific. This could apply also to the interface. Also, the name of the class no longer reflects its function, as it uses an arbitrary directory for the temp files, not just the working directory.

File: lang/cs/Org.Apache.REEF.Common/Io/WorkingDirectoryTempFileCreator.cs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        }
+
+        /// <summary>
+        /// Create a temp file sub folder 
+        /// </summary>
+        /// <param name="subFolder"></param>
+        /// <returns></returns>
+        public string CreateTempDirectory(string subFolder)
+        {
+            string folder = _tempFileFolder + subFolder;
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Io/WorkingDirectoryTempFileCreator.cs]: Please use `Path.Combine()`

File: lang/cs/Org.Apache.REEF.Common/Io/WorkingDirectoryTempFileCreator.cs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            if (!Directory.Exists(folder))
+            {
+                Directory.CreateDirectory(folder);
+            }
+
+            var path = Path.GetFullPath(folder);
+            Logger.Log(Level.Info, string.Format(CultureInfo.CurrentCulture, "Temp directory: {0} is craeted", path));
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Io/WorkingDirectoryTempFileCreator.cs]: Typo. Also, this probably shouldn't go into `Info`

File: lang/cs/Org.Apache.REEF.IO/FileSystem/WorkingDirectoryTempFileConfigModule.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+using Org.Apache.REEF.Common.Client.Parameters;
+using Org.Apache.REEF.Common.Io;
+using Org.Apache.REEF.Tang.Formats;
+using Org.Apache.REEF.Tang.Interface;
+using Org.Apache.REEF.Tang.Util;
+
+namespace Org.Apache.REEF.IO.FileSystem
+{
+    public sealed class WorkingDirectoryTempFileConfigModule : ConfigurationModuleBuilder
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO/FileSystem/WorkingDirectoryTempFileConfigModule.cs]: Document.

File: lang/cs/Org.Apache.REEF.IO/FileSystem/WorkingDirectoryTempFileConfigModule.cs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+ * under the License.
+ */
+
+using Org.Apache.REEF.Common.Client.Parameters;
+using Org.Apache.REEF.Common.Io;
+using Org.Apache.REEF.Tang.Formats;
+using Org.Apache.REEF.Tang.Interface;
+using Org.Apache.REEF.Tang.Util;
+
+namespace Org.Apache.REEF.IO.FileSystem
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO/FileSystem/WorkingDirectoryTempFileConfigModule.cs]: This has nothing to do with our distributed file system abstraction and should go to a different namespace.

File: lang/cs/Org.Apache.REEF.IO/FileSystem/WorkingDirectoryTempFileConfigModule.cs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+{
+    public sealed class WorkingDirectoryTempFileConfigModule : ConfigurationModuleBuilder
+    {
+        public static readonly OptionalParameter<string> TempFileFolerParameter = new OptionalParameter<string>();
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO/FileSystem/WorkingDirectoryTempFileConfigModule.cs]: Typo. Also, please document.

File: lang/cs/Org.Apache.REEF.Common/Io/ITempFileCreator.cs [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+using Org.Apache.REEF.Tang.Annotations;
+
+namespace Org.Apache.REEF.Common.Io
+{
+    /// <summary>
+    /// Provide interface for wraps temp file folder and temp file folder creation
+    /// </summary>
+    [DefaultImplementation(typeof(WorkingDirectoryTempFileCreator))]
+    public interface ITempFileCreator
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Io/ITempFileCreator.cs]: Please document the members of this interface.

File: lang/cs/Org.Apache.REEF.Common/Io/ITempFileCreator.cs [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+using Org.Apache.REEF.Tang.Annotations;
+
+namespace Org.Apache.REEF.Common.Io
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Io/ITempFileCreator.cs]: Why common.io, not just io?

File: lang/cs/Org.Apache.REEF.Common/Io/ITempFileCreator.cs [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+{
+    /// <summary>
+    /// Provide interface for wraps temp file folder and temp file folder creation
+    /// </summary>
+    [DefaultImplementation(typeof(WorkingDirectoryTempFileCreator))]
+    public interface ITempFileCreator
+    {
+        string TempFileFolder { get; }
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Io/ITempFileCreator.cs]: Wouldn't it be better to have methods that create new temp file names for the user? Maybe with configurable prefix and suffix and we put a GUID or such in the middle, making sure the file doesn't already exist.

File: lang/cs/Org.Apache.REEF.Common/Io/ITempFileCreator.cs [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+        string CreateTempDirectory(string subFolder);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Common/Io/ITempFileCreator.cs]: What if `subfolder` already exists? Same as above: Maybe give a method that creates the temp folder with a random, unique name and optional user-supplied prefix and suffix.

**Comment by markusweimer** [comment]
@jwang98052 I've done a pass.

File: lang/cs/Org.Apache.REEF.Common/Io/TempFileFolder.cs (hunk 1 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Io/TempFileFolder.cs]: done

File: lang/cs/Org.Apache.REEF.Common/Io/WorkingDirectoryTempFileCreator.cs (hunk 2 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Io/WorkingDirectoryTempFileCreator.cs]: done

File: lang/cs/Org.Apache.REEF.Common/Io/WorkingDirectoryTempFileCreator.cs (hunk 3 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Io/WorkingDirectoryTempFileCreator.cs]: done

File: lang/cs/Org.Apache.REEF.Common/Io/WorkingDirectoryTempFileCreator.cs (hunk 5 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Io/WorkingDirectoryTempFileCreator.cs]: done

File: lang/cs/Org.Apache.REEF.Common/Io/WorkingDirectoryTempFileCreator.cs (hunk 6 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Io/WorkingDirectoryTempFileCreator.cs]: done

File: lang/cs/Org.Apache.REEF.Common/Io/ITempFileCreator.cs (hunk 11 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Io/ITempFileCreator.cs]: TempFileCreator in Java is under reef-common.io. 
In C#, I can put all those including TempFileCreator. configuration module and provider, named parameter for temp folder:
1.under REEF.IO.TempFile
2. have TempFile folder under REEF.IO.FileSystem
What do you suggest @markusweimer ?

File: lang/cs/Org.Apache.REEF.Common/Io/ITempFileCreator.cs (hunk 13 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Io/ITempFileCreator.cs]: Good suggestion.

File: lang/cs/Org.Apache.REEF.Common/Io/ITempFileCreator.cs (hunk 12 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Io/ITempFileCreator.cs]: done

File: lang/cs/Org.Apache.REEF.Common/Io/ITempFileCreator.cs (hunk 10 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Common/Io/ITempFileCreator.cs]: done

File: lang/cs/Org.Apache.REEF.IO/FileSystem/WorkingDirectoryTempFileConfigModule.cs (hunk 9 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IO/FileSystem/WorkingDirectoryTempFileConfigModule.cs]: done

File: lang/cs/Org.Apache.REEF.IO/FileSystem/WorkingDirectoryTempFileConfigModule.cs (hunk 8 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IO/FileSystem/WorkingDirectoryTempFileConfigModule.cs]: create a separated folder under IO for it.

File: lang/cs/Org.Apache.REEF.IO/FileSystem/WorkingDirectoryTempFileConfigModule.cs (hunk 7 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IO/FileSystem/WorkingDirectoryTempFileConfigModule.cs]: done

**Comment by jwang98052** [comment]
@markusweimer thanks for the review! I have addressed all the review comments and pushed in to the branch.

**Comment by jwang98052** [comment]
@markusweimer anymore comments?

File: lang/cs/Org.Apache.REEF.IO/TempFileCreation/ITempFileCreator.cs [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    [DefaultImplementation(typeof(TempFileCreator))]
+    public interface ITempFileCreator
+    {
+        /// <summary>
+        /// return the temp folder path
+        /// </summary>
+        string TempFileFolder { get; }
+
+        /// <summary>
+        /// CReate a temp folder with given prefix and suffix as part of the name
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO/TempFileCreation/ITempFileCreator.cs]: Typo

File: lang/cs/Org.Apache.REEF.IO/TempFileCreation/ITempFileCreator.cs [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+        /// </summary>
+        /// <param name="prefix"></param>
+        /// <param name="suffix"></param>
+        /// <returns></returns>
+        string CreateTempDirectory(string prefix = "", string suffix = "");
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO/TempFileCreation/ITempFileCreator.cs]: We might want a similar method `GetTempFileName` which returns the absolute path to a temp file to be used. That way, we can exert some level of control over these in the future (e.g. spread them evenly across disks).

File: lang/cs/Org.Apache.REEF.IO/TempFileCreation/ITempFileCreator.cs [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+    /// <summary>
+    /// Provide interface to wrap temp file folder and temp file folder creation
+    /// </summary>
+    [DefaultImplementation(typeof(TempFileCreator))]
+    public interface ITempFileCreator
+    {
+        /// <summary>
+        /// return the temp folder path
+        /// </summary>
+        string TempFileFolder { get; }
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO/TempFileCreation/ITempFileCreator.cs]: Do we *need* to expose this, if we also add `GetTempFileName` as suggested below?

File: lang/cs/Org.Apache.REEF.IO/TempFileCreation/TempFileConfigurationModule.cs [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+using Org.Apache.REEF.Common.Client.Parameters;
+using Org.Apache.REEF.Tang.Formats;
+using Org.Apache.REEF.Tang.Interface;
+using Org.Apache.REEF.Tang.Util;
+
+namespace Org.Apache.REEF.IO.TempFileCreation
+{
+    public sealed class TempFileConfigurationModule : ConfigurationModuleBuilder
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO/TempFileCreation/TempFileConfigurationModule.cs]: Add documentation to this class and its members.

**Comment by markusweimer** [comment]
@jwang98052 I've done another pass.

File: lang/cs/Org.Apache.REEF.IO/TempFileCreation/ITempFileCreator.cs (hunk 14 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IO/TempFileCreation/ITempFileCreator.cs]: done

File: lang/cs/Org.Apache.REEF.IO/TempFileCreation/ITempFileCreator.cs (hunk 15 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IO/TempFileCreation/ITempFileCreator.cs]: The instance variable in this object is temp file folder rather than temp file name. The temp file folder is what we allow the user to set in configuration. With the same TempFileCreator, the user can call CreateTempDirectory to create different subfolders. So far, this interface/class is for temp folder not for a particular file. May be I should change the name to TempFIleFolderCreator?

File: lang/cs/Org.Apache.REEF.IO/TempFileCreation/ITempFileCreator.cs (hunk 16 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IO/TempFileCreation/ITempFileCreator.cs]: same as above

**Comment by jwang98052** [comment]
Addressed review comments and pushed in. 
@markusweimer see my comments regarding the file name and let me know your opinion.

File: lang/cs/Org.Apache.REEF.IO/TempFileCreation/ITempFileCreator.cs (hunk 15 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO/TempFileCreation/ITempFileCreator.cs]: I agree: In the current implementation of this interface, there is exactly one temp file folder.

However, I don't think the application author needs to know this. And we should be free to change this in the future. Nor is it necessary for the app author to choose the file name of their temp files and take care of their deletion after the fact.

If we choose the file names, we can also record them to be deleted at the end of the process, for example. Also, it would allow us to support non-traditional temp file locations in the future (e.g. round-robin on a couple of drives).

Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IO/TempFileCreation/ITempFileCreator.cs]: If you look at corresponding Java implementation in WorkingDirectoryTempFileCreator, the class wraps a folder (as File and as Path) rather than a file name. 
In createTempDirectory methods, user can specify prefix and sufix, the directory is still based on the temp folder. If the class wraps a file name, the createTempDirectory methods would make no sense. 
File deletion should be the responsibility of the one who calls a lib method to create the file. In C#, there are many ways to create a file, we don't want to fix a file creation in this class I think. It is really for a temp folder creation. 
Also if this class is to wraps a file name, that means the named parameter should be a file name instead of folder name, right?

Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO/TempFileCreation/ITempFileCreator.cs]: I think we are talking cross purpose. Let me try again :-)

What I am suggesting is a new method `string GetTempFileName(string prefix, string suffix)`. It returns the *absolute path* of a new temp file (that doesn't exist yet). Each time it is called, it returns a *new* file name. Skipping all sorts of defensive programming, a simple implementation could consist of:

```cs
public string GetTempFileName(string prefix, string suffix){
  return Path.GetFullPath(Path.Combine(_tempFileFolder, prefix+Guid.NewGuid().ToString("N")+suffix));
}
```

This is very likely the code users would write around this anyhow after calling `GetTempFileFolder` in this class.

File: lang/cs/Org.Apache.REEF.IO/TempFileCreation/TempFileCreator.cs [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+
+        /// <summary>
+        /// Create a unique subfolder under the _tempFileFolder with given prefix and suffix
+        /// </summary>
+        /// <param name="prefix"></param>
+        /// <param name="suffix"></param>
+        /// <returns></returns>
+        public string CreateTempDirectory(string prefix, string suffix)
+        {
+            string fullPathOfSubFolder = Path.GetFullPath(Path.Combine(_tempFileFolder, prefix, Guid.NewGuid().ToString("N").Substring(0, 8), suffix));
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.IO/TempFileCreation/TempFileCreator.cs]: If I understand the [documentation](https://msdn.microsoft.com/en-us/library/dd782933%28v=vs.110%29.aspx) on `Path.Combine()` correctly, this doesn't implement the desired behavior, as it creates a folder `_tempFileFolder\prefix\GUID\suffix`. Shouldn't this be `Path.GetFullPath(Path.Combine(_tempFileFolder, prefix+ Guid.NewGuid().ToString("N").Substring(0, 8)+ suffix)`? Or, better yet, something made via `string.format()` instead of the `+`?

File: lang/cs/Org.Apache.REEF.IO/TempFileCreation/ITempFileCreator.cs (hunk 15 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IO/TempFileCreation/ITempFileCreator.cs]: Got you this time :).

File: lang/cs/Org.Apache.REEF.IO/TempFileCreation/TempFileCreator.cs (hunk 18 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IO/TempFileCreation/TempFileCreator.cs]: Good catch. I might just use + as there is no other strings in between if using format.

**Comment by jwang98052** [comment]
@markusweimer have addressed review comments and pushed in. Thanks!

**Comment by markusweimer** [comment]
I'll test and merge.
